### PR TITLE
overlord: implement Get/Set/Copy on State

### PR DESCRIPTION
**Pull request by pedronis**
This implements accessors and copy on State.

Following the original sketch Get and Set don't return errors but panic, this seems the rational behaviour, there is no clear local recovery path at that point.

I think we will need testing, preference for backward compatibility and versioning approaches at the level of Read/WriteState and journaling to avoid getting into those panics while internal formats may evolve.

### CODE REVIEW

File: overlord/state.go [hunk 1]
```diff
@@ -20,25 +20,49 @@
...
+	}
+	s.entries[key] = serialized
 }
 
 // Copy returns an indepent copy of the state.
 func (s *State) Copy() *State {
-	return nil
+	entries := make(map[string]json.RawMessage, len(s.entries))
+	for k, s := range s.entries {
+		entries[k] = s
```
Comment (zyga) [inline overlord/state.go]: Should we .Copy() here as well? I don't know if RawMessage is mutable or not

**Comment by zyga** [comment]
+1 with one comment above.

File: overlord/state.go (hunk 1 above)
Comment (pedronis) [inline overlord/state.go]: it's mutable but we don't leak them outside, and the code only Umarshals them

File: overlord/state.go [hunk 2]
```diff
@@ -20,25 +20,49 @@
...
+		entries: make(map[string]json.RawMessage),
+	}
+}
 
 // Get unmarshals the stored value associated with the provided key
 // into the value parameter.
 func (s *State) Get(key string, value interface{}) {
+	err := json.Unmarshal(s.entries[key], value)
+	if err != nil {
+		panic(fmt.Errorf("internal error: could not retrieve and unmarshal state entry %q: %v", key, err))
```
Comment (chipaca) [inline overlord/state.go]: maybe use logger.Panicf?

File: overlord/state.go [hunk 3]
```diff
@@ -20,25 +20,49 @@
...
 // The provided value must properly marshal and unmarshal with encoding/json.
 func (s *State) Set(key string, value interface{}) {
+	serialized, err := json.Marshal(value)
+	if err != nil {
+		panic(fmt.Errorf("internal error: could not marshal value for state entry %q: %v", key, err))
+	}
+	s.entries[key] = serialized
 }
 
 // Copy returns an indepent copy of the state.
```
Comment (chipaca) [inline overlord/state.go]: independent?

**Comment by chipaca** [comment]
What's guarding state from concurrent access?

**Comment by chipaca** [comment]
:+1:

File: overlord/state.go (hunk 2 above)
Comment (niemeyer) [inline overlord/state.go]: I wonder if this should even be a panic. It seems like the outcome of something assuming facts about the state which change with future versions may be hard to get precisely right every time. Might be better for us to be less confident about our own ability to get it right every time, and instead have a slightly softer failure mode for when it happens.

File: overlord/state.go [hunk 4]
```diff
@@ -20,25 +20,49 @@
...
+		panic(fmt.Errorf("internal error: could not retrieve and unmarshal state entry %q: %v", key, err))
+	}
 }
 
 // Set associates value with key for future consulting by managers.
 // The provided value must properly marshal and unmarshal with encoding/json.
 func (s *State) Set(key string, value interface{}) {
+	serialized, err := json.Marshal(value)
+	if err != nil {
+		panic(fmt.Errorf("internal error: could not marshal value for state entry %q: %v", key, err))
```
Comment (niemeyer) [inline overlord/state.go]: This seems fine as a panic, though. If the thing in its present version cannot be marshaled as JSON, we have a larger problem, and definitely some missing tests.

File: overlord/state.go (hunk 2 above)
Comment (niemeyer) [inline overlord/state.go]: I suggest adding an error result here, with two possible outcomes: a clear JSON unmarshal issue, or ErrNoState, returned when the key is missing altogether.

**Comment by niemeyer** [comment]
LGTM with the suggested change.

**Comment by niemeyer** [comment]
Regarding concurrent access, it's unclear if we need it right now. If we do, we need to expose locking through its interface, so that call sites are comfortable that they may fiddle sequentially with the state reading and writing without it changing behind their back. So likely state.Lock/Unlock.

**Comment by pedronis** [comment]
my current thinking is that there should be one goroutine responsible for state manipulation, in a very strict erlang-like model, as long as we can keep a clear separation between long running (that we spawn out) vs normal/quick bits of Apply logic that should be cleanest/safest. One interesting thought though is that if we expect Apply to spawn out long running things, Apply impls cannot be idempotent in naive ways by starting from scratch, they need good ways to know that a subtask has been initiated, is still running with some good sanity checking in the mix

**Comment by pedronis** [comment]
retest this please

**Pull request merged**